### PR TITLE
Deduped API fields, expanded Workflows

### PR DESCRIPTION
**Pull request by copejon**

### CODE REVIEW

File: keps/sig-storage/20191125-bucket-provisioning.md [hunk 1]
```diff
@@ -131,75 +129,83 @@ Sidecar start up will follow these steps:
...
+| ----------------------------- | -------------- | ------------------ |
+| **objectBucketName: non-nil** | Brownfield     | Undefined          |
+| **objectBucketName: nil**     | Greenfield     | Static             |
+
+#### Create Bucket (Greenfield)
+
+1. The user creates a `Bucket` in their namespace, with reference to a `BucketClass`.
+1. The Controller sees the new `Bucket` and applies a `finalizer` for orchestrated deletions.
+1. The Controller gets the `BucketClass` referenced by the `Bucket`.
+1. The Controller creates a `BucketContent` object with its `BucketClassName` set to the name of its `BucketClass` and a `finalizer`.
```
Comment (jeffvance) [inline keps/sig-storage/20191125-bucket-provisioning.md]: nit: How about "...a `BucketContent` object with its `BucketClassName` set to the name of **the** `BucketClass`..."

and ditto below.

File: keps/sig-storage/20191125-bucket-provisioning.md [hunk 2]
```diff
@@ -131,75 +129,83 @@ Sidecar start up will follow these steps:
...
+1. The Sidecar detects the new `BucketContent` object and gets the associated `BucketClass`.
+1. The Sidecar calls the CreateBucket() rpc, passing the `bucketClass.parameters` and is returned a bucket endpoint and credentials.
+1. The Sidecar creates a secret containing the endpoint and credentials, with a random/unique name and `ownerRef` set to `BucketContent`.
+1. The Sidecar updates `BucketContent.secretRef` with its `secret` name and namespace and sets `BucketContent.status.phase` to *“Ready”.*
+1. The Controller detects the `BucketContent` update and sees the *“Ready”* phase. 
+1. The Controller copies the generated `secret` to the `Bucket` namespace with name defined in `Bucket.secretName`.  The `secret` is created with an `ownerRef` of the `Bucket`.
+1. The Controller “binds” the `Bucket` and `BucketContent` by setting `bucket.status.bucketContentName` and `bucketContent.bucketRef`, and sets both statuses to *“Bound”*.
+1. The app `pod` ingests `secret` and runs.
+
+#### Grant Bucket Access (Brownfield)
```
Comment (jeffvance) [inline keps/sig-storage/20191125-bucket-provisioning.md:156]: _Maybe_ greenfield and brownfield create can be combined (like for delete). They are very similar and combining helps illustrate our goal of keeping brownfield and greenfield workflows similar.

File: keps/sig-storage/20191125-bucket-provisioning.md [hunk 3]
```diff
@@ -254,33 +258,29 @@ Metadata:
   - cosi.io/finalizer [3]
 spec:
   bucketClassName: [4]
-  supportedProtocols: [5]
-  releasePolicy: {"Delete", "Retain"} [6]
-  bucketRef: [7]
+	bucketRef: [5]
```
Comment (jeffvance) [inline keps/sig-storage/20191125-bucket-provisioning.md]: leading tab is red

File: keps/sig-storage/20191125-bucket-provisioning.md [hunk 4]
```diff
@@ -254,33 +258,29 @@ Metadata:
...
     name:
     namespace:
-  secretName: [8]
-  accessMode: {"rw", "ro"} [9]
+  secretRef: [6]
```
Comment (jeffvance) [inline keps/sig-storage/20191125-bucket-provisioning.md:264]: do you want to call out `name:` and `namespace:` as done for `bucketRef`?

File: keps/sig-storage/20191125-bucket-provisioning.md [hunk 5]
```diff
@@ -254,33 +258,29 @@ Metadata:
...
+1. `name`: Generated in the pattern of `<BUCKET-CLASS-NAME>'-'<RANDOM-SUFFIX>`. 
+1. `labels`: COSI controller adds the label to its managed resources for easy CLI GET ops.  Value is the driver name returned by GetDriverInfo() rpc. Characters that do not adhere to [Kubernetes label conventions](https://kubernetes.io/docs/concepts/overview/working-with-objects/labels/#syntax-and-character-set) will be converted to ‘-’.
 1. `finalizers`: COSI controller adds the finalizer to defer Bucket deletion until backend deletion ops succeed.
 1. `bucketClassName`: Name of the associated `BucketClass`.
-1. `supportedProtocols`:  String array of protocols (e.g. s3, gcs, swift, etc.) supported by the associated object store.
-1. `releasePolicy`: the release policy defined in the associated BucketClass (see [BucketClass](#BucketClass) for more information).
 1. `bucketRef`: the name & namespace of the associated `Bucket`. For both brownfield cases, `bucketRef` is set by an admin and names the bucket to be accessed. For greenfield, it is added by the COSI controller.
```
Comment (jeffvance) [inline keps/sig-storage/20191125-bucket-provisioning.md]: Delete this:
~~For both brownfield cases, `bucketRef` is set by an admin and names the bucket to be accessed. For greenfield, it is added by the COSI controller.~~

File: keps/sig-storage/20191125-bucket-provisioning.md [hunk 6]
```diff
@@ -254,33 +258,29 @@ Metadata:
...
-1. `secretName`: the name of the app secret in the `Bucket`'s namespace, as define in `bucketRef.namespace`. For static brownfield, `secretName` is set by an admin; otherwise it is added by the COSI controller.
-1. `accessMode`: The level of access granted to the credentials stored in `secretRef`, one of "read only" or "read/write".
+1. `secretRef`: the name and namespace of the source secret in the `Bucket`'s namespace, as define in `bucketRef.namespace`.  This `secret` may be generated by the driver or written by an admin.
```
Comment (jeffvance) [inline keps/sig-storage/20191125-bucket-provisioning.md]: Maybe some rewording:
"`secretRef`: the name and namespace of the secret to be copied to the `Bucket`'s namespace.  This `secret` may be generated by the driver or created by an admin."

File: keps/sig-storage/20191125-bucket-provisioning.md [hunk 7]
```diff
@@ -296,25 +296,25 @@ metadata:
   name: 
 provisioner: [1]
 supportedProtocols: [2]
-accessMode: {"ro", "rw"} [3]
+accessMode: {"ro", "wo", "rw"} [3]
```
Comment (jeffvance) [inline keps/sig-storage/20191125-bucket-provisioning.md:300]: For the paragraph above, under ### BucketClass:
"...In dynamic brownfield workflows, the BucketClass contains a reference to a `objectBucketName` which names the existing object store bucket. In static brownfield workflows, the provisioner field is empty, the secret used to grant access to the bucket must be specified, and the `objectBucjketName` is unnecessary. In static brownfield, the secret must be created exactly as if generated by a driver, including endpoint, credentials, bucket name, etc.

There is currently no default bucket class."

File: keps/sig-storage/20191125-bucket-provisioning.md [hunk 8]
```diff
@@ -296,25 +296,25 @@ metadata:
...
-1. `supportedProtocols`: (Optional) A strings array of protocols the associated object store supports (e.g. swift, s3, gcs, etc.). If empty then protocol checking is skipped.
-1. `accessModes`: (Optional) Declares the level of access given to credentials provisioned through this class. If empty then "rw" is assumed.
-1.  `releasePolicy`: Prescribes outcome of a Delete events. **Note:** `releasePolicy` is ignored for all brownfield cases.
+1. `provisioner`: (Optional) The name of the driver. If supplied the driver container and sidecar container are expected to be deployed. If omitted the `secretRef` is required for static provisioning.
+1. `supportedProtocols`: (Optional) An array of protocols the associated object store supports (e.g. swift, s3, gcs, etc.). *Only* serves a descriptive purpose and is not verified.
+1. `accessModes`: (Optional) Declares the level of access given to credentials provisioned through this class.     If empty, drivers may set defaults.
+1.  `releasePolicy`: Prescribes outcome of a Delete events. **Note:** In Brownfield and Static cases, *Retain* is mandated.
```
Comment (jeffvance) [inline keps/sig-storage/20191125-bucket-provisioning.md:312]: what is the default?

File: keps/sig-storage/20191125-bucket-provisioning.md (hunk 2 above)
Comment (copejon) [reply keps/sig-storage/20191125-bucket-provisioning.md:156]: I agree, but I think we can leave this until after the meeting.

File: keps/sig-storage/20191125-bucket-provisioning.md (hunk 4 above)
Comment (copejon) [reply keps/sig-storage/20191125-bucket-provisioning.md]: Not _really_ as they're already core pieces of k8s.

**Pull request merged**